lib/UR/Role.pm

```diff
@@ -0,0 +1,732 @@
+                 . ';';
+
+    my $exception;
+    do {
+        local $@;
+        eval $string;
+        $exception = $@;
+    };
+
+    if ($exception) {
+        Carp::croak("Failed to apply overloads to package $class_name: $exception");
+    }
+    return 1;
+}
+
+sub _define_role {
+    my($role_name, $func, @params) = @_;
+
+    if (defined($func) and $func eq "role" and @params > 1 and $role_name ne "UR::Role") {
+        my @role_params;
+        if (@params == 2 and ref($params[1]) eq 'HASH') {
+            @role_params = %{ $params[1] };
+        }
+        elsif (@params == 2 and ref($params[1]) eq 'ARRAY') {
+            @role_params = @{ $params[1] };
+        }
+        else {
+            @role_params = @params[1..$#params];
+        }
+        my $role = UR::Role->define(role_name => $role_name, @role_params);
+        unless ($role) {
+            Carp::croak "error defining role $role_name!";
+        }
+        return sub { $role_name };
+    } else {
+        return;
+    }
+}
+
+sub defer($) {
+    Carp::croak('defer takes only a single argument') unless (@_ == 1);
+    return UR::Role::DeferredValue->create(id => shift);
+}
+
+1;
+
+__END__
+
+=pod
+
+=head1 NAME
+
+UR::Role - Roles in UR, an alternative to inheritance
+
+=head1 SYNOPSIS
+
+  package My::Role;
+  role My::Role {
+      has => [
+          role_property => { is => 'String' },
+          another_prop  => { is => 'Integer' },
+      },
+      requires => ['class_method'],
+      excludes => ['Bad::Role'],
+  };
+  sub role_method { ... }
+
+
+  package My::Class;
+  class My::Class {
+      has => [
+          class_property => { is => 'Integer ' },
+      ],
+      roles => ['My::Role'],
+  };
+  sub class_method { ... }
+
+  my $obj = My::Class->new();
+  $obj->does('My::Role');  # true
+
+=head1 DESCRIPTION
+
+Roles are used to encapsulate a piece of behavior to be used in other classes.
+They have properties and methods that get melded into any class that composes
+them.  A Role can require any composing class to implement a list of methods
+or properties.
+
+Roles are not classes.  They can not be instantiated or inherited from.  They
+are composed into a class by listing their names in the C<roles> attribute of
+a class definition.
+
+=head2 Defining a Role
+
+Roles are defined with the C<role> keyword.  Their definition looks very
+similar to a class definition as described in L<UR::Object::Type::Initializer>.
+In particular, Roles have a C<has> section to define properties, and accept
+many class-meta attributes such as 'id_generator', 'valid_signals', and 'doc'.
+
+Roles may implement operator overloading via the 'use overload' mechanism.
+
+Roles also have unique atributes to declare restrictions on their use.
+
+=over 4
+
+=item requires
+
+A listref of property and method names that must appear in any class composing
+the Role.  Properties and methods defined in other roles or parent classes
+can satisfy a requirement.
+
+=item excludes
```

Comment:
I don't think it's necessary, but I'll add it in if you feel strongly about it.  Saying it's for "incompatibilities" seems pretty clear about it's use.